data/src/update.rs

```diff
@@ -19,24 +21,107 @@ impl defmt::Format for UpdateKey {
 
 /// A constrained form of pre-release designators along with
 /// a numeric identifer.
-#[derive(Copy, Clone, Eq, PartialEq, Deserialize, Serialize)]
-#[non_exhaustive]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub enum PreRelease {
-    Alpha(u32),
-    Beta(u32),
+    Alpha(u8),
+    Beta(u8),
+}
+impl Ord for PreRelease {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self, other) {
+            (PreRelease::Alpha(self_ident), PreRelease::Alpha(other_ident)) => {
+                self_ident.cmp(other_ident)
+            }
+            (PreRelease::Alpha(_), PreRelease::Beta(_)) => Ordering::Less,
+            (PreRelease::Beta(_), PreRelease::Alpha(_)) => Ordering::Greater,
+            (PreRelease::Beta(self_ident), PreRelease::Beta(other_ident)) => {
+                self_ident.cmp(other_ident)
+            }
+        }
+    }
+}
+impl PartialOrd for PreRelease {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
 }
 
 /// A compact and limited representation of a version based on
 /// https://semver.org. In particular, there is no provision for a
 /// build identifier. Also, pre-releases are constrained to Alpha
-/// and Beta.
-#[derive(Clone, Eq, PartialEq, Deserialize, Serialize)]
+/// and Beta and must always have an ident.
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub struct Version {
     pub major: u8,
     pub minor: u8,
     pub patch: u8,
     pub pre: Option<PreRelease>,
 }
+#[derive(Debug)]
+pub struct ParseVersionErr;
+impl FromStr for Version {
+    type Err = ParseVersionErr;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let Some((l, r)) = s.split_once('.') else {return Err(ParseVersionErr);};
```

Comment:
You could rewrite all these early return lines like this:

```
let (l, r) = s.split_once('.').ok_or(ParseVersionErr)?;
```

Reply:
Much better